analysis/topology/sequences.lean

```diff
@@ -0,0 +1,278 @@
+
+/-
+Copyright (c) 2018 Jan-David Salchow. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jan-David Salchow
+-/
+import analysis.topology.topological_space
+import analysis.metric_space
+import analysis.topology.uniform_space
+import data.set.countable
+import data.real.cau_seq_filter
+
+open filter
+
+/- Sequences in topological spaces.
+ - 
+ - In this file we define sequences in topological spaces and show how they are related to
+ - filters and the topology. In particular, we
+ -
+ - (*) associate a filter with a sequence and prove equivalence of convergence of the two,
+ - (*) define the sequential closure of a set and prove that it's contained in the closure, 
+ - (*) define a type class "sequential_space" in which closure and sequential closure agree, 
+ - (*) define sequential continuity and show that it coincides with continuity in sequential spaces,
+ - (*) provide an instance that shows that every metric space is a sequential space.
+ -
+ - TODO: There should be an instance that associates a sequential space with a first countable
+ -       space. 
+ - TODO: Sequential compactness should be handled here.
+ -/
+namespace sequence
+
+universes u v
+variables {X : Type u} {Y : Type v}
+
+def to_filter (x : ℕ → X) : filter X := filter.map x at_top
+
+/- Statements about sequences in general topological spaces. -/
+section topological_space
+variables [topological_space X] [topological_space Y]
+
+/-- The notion of convergence of sequences in topological spaces. -/
+def converges_to (x : ℕ → X) (limit : X) : Prop :=
+∀ U : set X, limit ∈ U → is_open U → ∃ n0 : ℕ, ∀ n ≥ n0, (x n) ∈ U
+
+lemma const_seq_conv (p : X) : converges_to (λ n, p) p :=
+assume U (_ : p ∈ U) (_ : is_open U), exists.intro 0 (assume n (_ : n ≥ 0), ‹p ∈ U›)
+
+/-- A sequence converges if and only if the associated statement for filter holds. -/
+lemma converges_to_iff_tendsto [topological_space X] {x : ℕ → X} {limit : X} :
+  converges_to x limit ↔ tendsto x at_top (nhds limit) := 
+iff.intro
+  (assume xtol : converges_to x limit,
+    suffices ∀ U, limit ∈ U → is_open U → x ⁻¹' U ∈ at_top.sets, from tendsto_nhds this,
+      assume U limitInU isOpenU,
+        suffices ∃ n0 : ℕ, ∀ n ≥ n0, (x n) ∈ U, by simp[this],
+        xtol U limitInU isOpenU)
+  (assume ttol : tendsto x at_top (nhds limit),
+    show ∀ U : set X, limit ∈ U → is_open U → ∃ n0 : ℕ, ∀ n ≥ n0, (x n) ∈ U, from
+      assume U limitInU isOpenU,
+        have {n | (x n) ∈ U} ∈ at_top.sets, from mem_map.mp $ le_def.mp ttol U 
+                                             (mem_nhds_sets isOpenU limitInU),
+        show ∃ n0 : ℕ, ∀ n ≥ n0, (x n) ∈ U, from mem_at_top_sets.mp this)
+
+/--
```

Comment:
remove the ` - ` in the next two lines. They occur in the tooltip, i.e, the tooltip is now `- The sequential closure of a subset M ⊆ X of a topological space X is - the set of all p ∈ X which arise as limit of sequences in M.`